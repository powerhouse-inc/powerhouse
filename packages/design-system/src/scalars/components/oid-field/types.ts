import type { OIDInputProps } from "../../../ui/components/data-entry/oid-input/types.js";
import type { FieldErrorHandling } from "../types.js";

type OIDFieldProps = OIDInputProps & FieldErrorHandling;

export type { OIDFieldProps };
