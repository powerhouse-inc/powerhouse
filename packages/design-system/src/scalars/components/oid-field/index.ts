export { OIDField } from "./oid-field.js";
export type { OIDFieldProps } from "./types.js";
