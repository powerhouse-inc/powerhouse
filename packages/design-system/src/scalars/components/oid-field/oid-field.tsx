import { OIDInput } from "../../../ui/components/data-entry/oid-input/index.js";
import { withFieldValidation } from "../fragments/with-field-validation/index.js";
import type { OIDFieldProps } from "./types.js";

const OIDField = withFieldValidation<OIDFieldProps>(OIDInput, {
  validations: {
    _validOIDFormat: () => (value: string | undefined) => {
      if (value === "" || value === undefined) {
        return true;
      }

      const uuidPattern =
        "^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$";

      const isValidUUID = new RegExp(uuidPattern).test(value);
      if (!isValidUUID) {
        return "Invalid uuid format. Please enter a valid uuid v4.";
      }

      return true;
    },
  },
});

OIDField.displayName = "OIDField";

export { OIDField };
