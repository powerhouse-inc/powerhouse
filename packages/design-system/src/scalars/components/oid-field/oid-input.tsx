/* eslint-disable react/jsx-props-no-spreading */
import React, { useCallback, useId } from "react";
import { IdAutocompleteListOption } from "../fragments/id-autocomplete/id-autocomplete-list-option.js";
import { IdAutocomplete } from "../fragments/id-autocomplete/index.js";
import type { OIDInputProps, OIDOption } from "./types.js";

const OIDInput = React.forwardRef<HTMLInputElement, OIDInputProps>(
  (
    {
      id: idProp,
      name,
      className,
      label,
      description,
      value,
      defaultValue,
      disabled,
      placeholder,
      required,
      errors,
      warnings,
      onChange,
      onBlur,
      onClick,
      onMouseDown,
      autoComplete: autoCompleteProp,
      variant = "withValue",
      fetchOptionsCallback,
      fetchSelectedOptionCallback,
      isOpenByDefault, // to be used only in stories
      initialOptions, // to be used only in stories
      ...props
    },
    ref,
  ) => {
    const prefix = useId();
    const id = idProp ?? `${prefix}-oid`;
    const autoComplete = autoCompleteProp ?? true;

    const renderOption = useCallback(
      (
        option: OIDOption,
        displayProps?: {
          asPlaceholder?: boolean;
          showValue?: boolean;
          isLoadingSelectedOption?: boolean;
          handleFetchSelectedOption?: (value: string) => void;
          isFetchSelectedOptionSync?: boolean;
          className?: string;
        },
      ) => (
        <IdAutocompleteListOption
          variant={variant}
          icon={option.icon}
          title={option.title}
          path={
            displayProps?.asPlaceholder
              ? { text: "Type not available" }
              : option.path
          }
          value={
            displayProps?.asPlaceholder ? "oid not available" : option.value
          }
          description={option.description}
          placeholderIcon="Braces"
          {...displayProps}
        />
      ),
      [variant],
    );

    return autoComplete && fetchOptionsCallback ? (
      <IdAutocomplete
        id={id}
        name={name}
        className={className}
        label={label}
        description={description}
        value={value}
        defaultValue={defaultValue}
        disabled={disabled}
        placeholder={placeholder}
        required={required}
        errors={errors}
        warnings={warnings}
        onChange={onChange}
        onBlur={onBlur}
        onClick={onClick}
        onMouseDown={onMouseDown}
        autoComplete={true}
        variant={variant}
        fetchOptionsCallback={fetchOptionsCallback}
        fetchSelectedOptionCallback={fetchSelectedOptionCallback}
        isOpenByDefault={isOpenByDefault}
        initialOptions={initialOptions}
        renderOption={renderOption}
        {...props}
        ref={ref}
      />
    ) : (
      <IdAutocomplete
        id={id}
        name={name}
        className={className}
        label={label}
        description={description}
        value={value}
        defaultValue={defaultValue}
        disabled={disabled}
        placeholder={placeholder}
        required={required}
        errors={errors}
        warnings={warnings}
        onChange={onChange}
        onBlur={onBlur}
        onClick={onClick}
        onMouseDown={onMouseDown}
        autoComplete={false}
        {...props}
        ref={ref}
      />
    );
  },
);

OIDInput.displayName = "OIDInput";

export { OIDInput };
