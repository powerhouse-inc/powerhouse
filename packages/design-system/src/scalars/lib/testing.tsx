import { render, type RenderResult } from "@testing-library/react";
import { Form } from "../components/index.js";

export const renderWithForm = (
  children: React.ReactNode,
  onSubmit?: () => void,
): RenderResult => {
  return render(children, {
    wrapper: ({ children }) => (
      <Form onSubmit={onSubmit ?? (() => null)}>{children}</Form>
    ),
  });
};
