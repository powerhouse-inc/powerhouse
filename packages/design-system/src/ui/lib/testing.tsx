import { Form } from "@powerhousedao/design-system";
import type { RenderResult } from "@testing-library/react";
import { render } from "@testing-library/react";

export const renderWithForm = (
  children: React.ReactNode,
  onSubmit?: () => void,
): RenderResult => {
  return render(children, {
    wrapper: ({ children }) => (
      <Form onSubmit={onSubmit ?? (() => null)}>{children}</Form>
    ),
  });
};
