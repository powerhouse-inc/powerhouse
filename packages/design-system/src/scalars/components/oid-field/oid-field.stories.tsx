import type { Meta, StoryObj } from "@storybook/react";
import {
  fetchOptions,
  fetchSelectedOption,
  mockedOptions,
} from "../../../ui/components/data-entry/oid-input/mocks.js";
import { withForm } from "../../lib/decorators.js";
import {
  getDefaultArgTypes,
  getValidationArgTypes,
  PrebuiltArgTypes,
  StorybookControlCategory,
} from "../../lib/storybook-arg-types.js";
import { OIDField } from "./oid-field.js";

const meta: Meta<typeof OIDField> = {
  title: "Document Engineering/Scalars/OID Field",
  component: OIDField,
  decorators: [
    withForm,
    (Story) => (
      <div style={{ maxWidth: "280px", margin: "1rem auto 0" }}>
        <Story />
      </div>
    ),
  ],
  parameters: {
    layout: "padded",
  },
  tags: ["autodocs"],
  argTypes: {
    ...getDefaultArgTypes(),
    ...PrebuiltArgTypes.placeholder,

    autoComplete: {
      control: "boolean",
      description:
        "Enables autocomplete functionality to suggest options while typing",
      table: {
        type: { summary: "boolean" },
        defaultValue: { summary: "true" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },

    fetchOptionsCallback: {
      control: "object",
      description:
        "Function to fetch options based on user input and context. " +
        "Must return a Promise that resolves to an array of objects or an array of objects with the following properties:\n\n" +
        "icon?: IconName | React.ReactElement\n\n" +
        "title?: string\n\n" +
        "path?: string | { text: string; url: string; }\n\n" +
        "value: string\n\n" +
        "description?: string\n\n",
      table: {
        type: {
          summary:
            "(userInput: string; context?: {}) => Promise<OIDOption[]> | OIDOption[]",
        },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
        readonly: true,
      },
      if: { arg: "autoComplete", neq: false },
    },

    fetchSelectedOptionCallback: {
      control: "object",
      description:
        "Function to fetch details for a selected option. " +
        "Must return a Promise that resolves to an object or an object with the following properties:\n\n" +
        "icon?: IconName | React.ReactElement\n\n" +
        "title?: string\n\n" +
        "path?: string | { text: string; url: string; }\n\n" +
        "value: string\n\n" +
        "description?: string\n\n" +
        "or undefined if the option is not found",
      table: {
        type: {
          summary:
            "(value: string) => Promise<OIDOption | undefined> | OIDOption | undefined",
        },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
        readonly: true,
      },
      if: { arg: "autoComplete", neq: false },
    },

    previewPlaceholder: {
      control: "object",
      description:
        "Custom placeholder values to show when no option is selected or when there are no matching options. " +
        "Can include custom values for:\n\n" +
        "icon?: IconName | React.ReactElement\n\n" +
        "title?: string\n\n" +
        "path?: string | { text: string; url: string; }\n\n" +
        "value: string\n\n" +
        "description?: string\n\n",
      table: {
        type: { summary: "OIDOption" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
      if: { arg: "autoComplete", neq: false },
    },

    variant: {
      control: "radio",
      options: [
        "withValue",
        "withValueAndTitle",
        "withValueTitleAndDescription",
      ],
      description:
        "Controls the amount of information displayed for each option: value only, value with title, or value with title and description",
      table: {
        type: {
          summary:
            '"withValue" | "withValueAndTitle" | "withValueTitleAndDescription"',
        },
        defaultValue: { summary: "withValue" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
      if: { arg: "autoComplete", neq: false },
    },

    ...getValidationArgTypes(),
  },
  args: {
    name: "oid-field",
  },
} satisfies Meta<typeof OIDField>;

export default meta;

type Story = StoryObj<typeof OIDField>;

export const Default: Story = {
  args: {
    label: "OID field",
    placeholder: "uuid",
    fetchOptionsCallback: fetchOptions,
    fetchSelectedOptionCallback: fetchSelectedOption,
  },
};

export const Empty: Story = {
  args: {
    label: "OID field",
    placeholder: "uuid",
    isOpenByDefault: true,
    defaultValue: "uuid",
    variant: "withValueTitleAndDescription",
    fetchOptionsCallback: fetchOptions,
    fetchSelectedOptionCallback: fetchSelectedOption,
  },
};

export const Open: Story = {
  args: {
    label: "OID field",
    placeholder: "uuid",
    isOpenByDefault: true,
    defaultValue: "uuid",
    variant: "withValueTitleAndDescription",
    initialOptions: mockedOptions,
    fetchOptionsCallback: fetchOptions,
    fetchSelectedOptionCallback: fetchSelectedOption,
  },
};

export const Filled: Story = {
  args: {
    label: "OID field",
    placeholder: "uuid",
    defaultValue: mockedOptions[0].value,
    initialOptions: mockedOptions,
    variant: "withValueTitleAndDescription",
    fetchOptionsCallback: fetchOptions,
    fetchSelectedOptionCallback: fetchSelectedOption,
  },
};
