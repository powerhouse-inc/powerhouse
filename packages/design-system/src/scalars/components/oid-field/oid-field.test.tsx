import { render, screen, waitFor } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import { describe, expect, it, vi } from "vitest";
import { renderWithForm } from "../../lib/testing.js";
import { Form } from "../form/index.js";
import { OIDField } from "./oid-field.js";

describe("OIDField Component", () => {
  window.HTMLElement.prototype.scrollIntoView = vi.fn();
  window.Element.prototype.scrollTo = vi.fn();
  window.matchMedia = vi.fn().mockImplementation((query) => ({
    matches: false,
    media: query as string,
    onchange: null,
    addListener: vi.fn(),
    removeListener: vi.fn(),
    addEventListener: vi.fn(),
    removeEventListener: vi.fn(),
    dispatchEvent: vi.fn(),
  }));

  const mockedOptions = [
    {
      icon: "Braces",
      title: "Object A",
      path: "rwa-portfolio-a",
      value: "baefc2a4-f9a0-4950-8161-fd8d8cc7dea7",
      description: "Object A description",
    },
    {
      icon: "Braces",
      title: "Object B",
      path: "rwa-portfolio-b",
      value: "baefc2a4-f9a0-4950-8161-fd8d8cc6cdb8",
      description: "Object B description",
    },
  ];

  const defaultGetOptions = vi.fn().mockResolvedValue(mockedOptions);
  const defaultGetSelectedOption = vi
    .fn()
    .mockImplementation((value: string) => {
      return mockedOptions.find((option) => option.value === value);
    });

  it("should match snapshot", () => {
    const { asFragment } = renderWithForm(
      <OIDField
        name="oid"
        label="OID Field"
        placeholder="uuid"
        fetchOptionsCallback={defaultGetOptions}
        fetchSelectedOptionCallback={defaultGetSelectedOption}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render with label", () => {
    renderWithForm(
      <OIDField
        name="oid"
        label="Test Label"
        fetchOptionsCallback={defaultGetOptions}
        fetchSelectedOptionCallback={defaultGetSelectedOption}
      />,
    );
    expect(screen.getByText("Test Label")).toBeInTheDocument();
  });

  it("should render with description", () => {
    renderWithForm(
      <OIDField
        name="oid"
        label="Test Label"
        description="Test Description"
        fetchOptionsCallback={defaultGetOptions}
        fetchSelectedOptionCallback={defaultGetSelectedOption}
      />,
    );
    expect(screen.getByText("Test Description")).toBeInTheDocument();
  });

  it("should handle disabled state", () => {
    renderWithForm(
      <OIDField
        name="oid"
        label="Test Label"
        disabled
        fetchOptionsCallback={defaultGetOptions}
        fetchSelectedOptionCallback={defaultGetSelectedOption}
      />,
    );
    expect(screen.getByRole("combobox")).toBeDisabled();
  });

  it("should display error messages", async () => {
    renderWithForm(
      <OIDField
        name="oid"
        label="Test Label"
        errors={["Invalid OID format"]}
        fetchOptionsCallback={defaultGetOptions}
        fetchSelectedOptionCallback={defaultGetSelectedOption}
      />,
    );
    await waitFor(() =>
      expect(screen.getByText("Invalid OID format")).toBeInTheDocument(),
    );
  });

  it("should display warning messages", () => {
    renderWithForm(
      <OIDField
        name="oid"
        label="Test Label"
        warnings={["OID may be deprecated"]}
        fetchOptionsCallback={defaultGetOptions}
        fetchSelectedOptionCallback={defaultGetSelectedOption}
      />,
    );
    expect(screen.getByText("OID may be deprecated")).toBeInTheDocument();
  });

  it("should show autocomplete options", async () => {
    const user = userEvent.setup();
    const originalRandom = Math.random;
    Math.random = vi.fn().mockReturnValue(1);

    renderWithForm(
      <OIDField
        name="oid"
        label="Test Label"
        variant="withValueTitleAndDescription"
        fetchOptionsCallback={defaultGetOptions}
        fetchSelectedOptionCallback={defaultGetSelectedOption}
      />,
    );

    const input = screen.getByRole("combobox");
    await user.click(input);
    await user.clear(input);
    await user.type(input, "test");

    await waitFor(() => {
      expect(defaultGetOptions).toHaveBeenCalledWith("test", {});
    });

    await waitFor(() => {
      expect(input).toHaveAttribute("aria-expanded", "true");
      expect(screen.getByText(mockedOptions[0].title)).toBeInTheDocument();
      expect(screen.getByText(mockedOptions[1].title)).toBeInTheDocument();
    });

    Math.random = originalRandom;
  });

  it("should have correct ARIA attributes", async () => {
    renderWithForm(
      <OIDField
        name="oid"
        label="Test Label"
        required
        errors={["Error message"]}
        fetchOptionsCallback={defaultGetOptions}
        fetchSelectedOptionCallback={defaultGetSelectedOption}
      />,
    );

    const input = screen.getByRole("combobox");
    expect(input).toHaveAttribute("aria-required", "true");
    await waitFor(() => expect(input).toHaveAttribute("aria-invalid", "true"));
    expect(input).toHaveAttribute("aria-expanded", "false");
  });

  it("should show correct placeholders for different variants", () => {
    const { rerender } = renderWithForm(
      <OIDField
        name="oid"
        label="Test Label"
        variant="withValueTitleAndDescription"
        fetchOptionsCallback={defaultGetOptions}
        fetchSelectedOptionCallback={defaultGetSelectedOption}
      />,
    );

    expect(screen.getByText("Title not available")).toBeInTheDocument();
    expect(screen.getByText("Path not available")).toBeInTheDocument();
    expect(screen.getByText("Description not available")).toBeInTheDocument();

    rerender(
      <OIDField
        name="oid"
        label="Test Label"
        variant="withValueAndTitle"
        fetchOptionsCallback={defaultGetOptions}
        fetchSelectedOptionCallback={defaultGetSelectedOption}
      />,
    );

    expect(screen.getByText("Title not available")).toBeInTheDocument();
    expect(screen.getByText("Path not available")).toBeInTheDocument();
    expect(
      screen.queryByText("Description not available"),
    ).not.toBeInTheDocument();

    rerender(
      <OIDField
        name="oid"
        label="Test Label"
        variant="withValue"
        fetchOptionsCallback={defaultGetOptions}
        fetchSelectedOptionCallback={defaultGetSelectedOption}
      />,
    );

    expect(screen.queryByText("Title not available")).not.toBeInTheDocument();
    expect(screen.queryByText("Path not available")).not.toBeInTheDocument();
    expect(
      screen.queryByText("Description not available"),
    ).not.toBeInTheDocument();
  });

  it("should handle autoComplete disabled", () => {
    renderWithForm(
      <OIDField name="oid" label="Test Label" autoComplete={false} />,
    );

    expect(screen.queryByRole("combobox")).not.toBeInTheDocument();
    expect(screen.getByRole("textbox")).toBeInTheDocument();
  });

  it("should validate OID format on submit", async () => {
    const mockOnSubmit = vi.fn();
    const user = userEvent.setup();
    const validOID = mockedOptions[1].value;
    const invalidOID = "invalid-oid";

    render(
      <Form onSubmit={mockOnSubmit}>
        <OIDField
          name="oid"
          label="Test Label"
          fetchOptionsCallback={defaultGetOptions}
          fetchSelectedOptionCallback={defaultGetSelectedOption}
        />
        <button type="submit">Submit</button>
      </Form>,
    );

    const input = screen.getByRole("combobox");
    await user.click(input);
    await user.type(input, invalidOID);

    await user.click(screen.getByText("Submit"));
    await waitFor(() => {
      expect(mockOnSubmit).not.toHaveBeenCalled();
      expect(screen.getByText(/Invalid OID format/)).toBeInTheDocument();
    });

    await user.clear(input);
    await user.type(input, validOID);

    await user.click(screen.getByText("Submit"));
    await waitFor(() => {
      expect(mockOnSubmit).toHaveBeenCalledWith({
        oid: validOID,
      });
    });
  });

  it("should handle value changes and auto selection", async () => {
    const user = userEvent.setup();
    const originalRandom = Math.random;
    Math.random = vi.fn().mockReturnValue(1);

    renderWithForm(
      <OIDField
        name="oid"
        label="Test Label"
        variant="withValueTitleAndDescription"
        fetchOptionsCallback={defaultGetOptions}
        fetchSelectedOptionCallback={defaultGetSelectedOption}
      />,
    );

    const input = screen.getByRole("combobox");
    await user.click(input);
    await user.type(input, mockedOptions[0].value);

    await waitFor(() => {
      expect(input).toHaveAttribute("aria-expanded", "false");
    });

    await waitFor(() => {
      expect(screen.getByText(mockedOptions[0].title)).toBeInTheDocument();
      expect(screen.getByText(mockedOptions[0].path)).toBeInTheDocument();
      expect(
        screen.getByText(mockedOptions[0].description),
      ).toBeInTheDocument();
    });

    Math.random = originalRandom;
  });

  it("should not invoke onChange on mount when it has a defaultValue", () => {
    const onChange = vi.fn();
    renderWithForm(
      <OIDField
        name="oid"
        label="Test Label"
        defaultValue={mockedOptions[0].value}
        fetchOptionsCallback={defaultGetOptions}
        fetchSelectedOptionCallback={defaultGetSelectedOption}
        onChange={onChange}
      />,
    );
    expect(onChange).not.toHaveBeenCalled();
  });
});
