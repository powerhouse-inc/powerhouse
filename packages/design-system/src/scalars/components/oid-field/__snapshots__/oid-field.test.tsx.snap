// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OIDField Component > should match snapshot 1`] = `
<DocumentFragment>
  <form
    id=":r0:"
    novalidate=""
  >
    <div
      class="flex flex-col gap-2"
    >
      <label
        class="inline-flex items-center text-sm font-medium leading-4 text-gray-900 dark:text-gray-50 mb-[3px]"
        for=":r1:-oid"
      >
        OID Field
      </label>
      <div
        class="flex size-full flex-col rounded-md [&_[cmdk-label]]:hidden dark:bg-charcoal-900 bg-gray-100"
        cmdk-root=""
        tabindex="-1"
      >
        <label
          cmdk-label=""
          for="radix-:r6:"
          id="radix-:r5:"
          style="position: absolute; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0, 0, 0, 0); white-space: nowrap; border-width: 0px;"
        />
        <div
          class="group relative"
        >
          <input
            aria-autocomplete="list"
            aria-controls="radix-:r4:"
            aria-expanded="false"
            aria-invalid="false"
            aria-labelledby="radix-:r5:"
            autocomplete="off"
            autocorrect="off"
            class="flex h-9 w-full rounded-md text-sm font-normal leading-5 text-gray-900 dark:text-gray-50 dark:border-charcoal-700 dark:bg-charcoal-900 border border-gray-300 bg-white px-3 py-2 font-sans placeholder:text-gray-500 dark:placeholder:text-gray-600 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-gray-900 focus-visible:ring-offset-0 focus-visible:ring-offset-white dark:focus-visible:ring-charcoal-300 dark:focus-visible:ring-offset-charcoal-900 dark:focus:bg-charcoal-900 focus:bg-gray-50 disabled:cursor-not-allowed disabled:border-gray-300 disabled:bg-gray-50 disabled:text-gray-700 disabled:dark:border-charcoal-800 disabled:dark:bg-charcoal-900 disabled:dark:text-gray-300 pr-9"
            cmdk-input=""
            id=":r1:-oid"
            name="oid"
            placeholder="uuid"
            role="combobox"
            spellcheck="false"
            type="text"
            value=""
          />
          <div
            class="absolute right-3 top-1/2 flex size-4 -translate-y-1/2 items-center pointer-events-none"
          />
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
